### ISSUE TITLE
See references show unrelated results

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/sdk#56984]
In my screenshot I'll sepparate this repro into three files so the VSCode UI helps a bit to see the problem.

```dart
class A {
  void m1() {}
  void m2() {
    m1();
  }
}

class B extends A {
  @override
  void m2() {
    m1();
  }

  @override
  void m1() {
    print('B.m1');
  }
}

void f1(B m) {
  m
    ..m2()
    ..m1();
}

class C extends A {
  @override
  void m1 () {}
  @override
  void m2() {
    m1();
  }
}
```

Here you can see that in the above structure `B` and `C` are subclasses of `A`. In this example, `m1` is called under `A.m2`, `B.m2`, `C.m2` and under `f1`.

If you look for references of `A.m1` you get all calls to it in the list.

![image](https://github.com/user-attachments/assets/86d0ee6e-e7fe-4bb1-b9b6-fa0c37d54910)

Now the problem; if you ask for references for `B.m1` or `C.m1` you get the same unrelated results:

![image](https://github.com/user-attachments/assets/1cb64e78-b27a-41c9-a778-30fed51f9a84)

Here you can see that when asking for references of `C.m1` I get instance calls that would **_never_** be `C`. Inside `B` and `f1` in this case.

The same happens if you ask for references of `B.m1`, you get the call under `C.m2` which could never be a `B` instance. 

---

I'd like to ask for this to be worked on so that only `B` instances (or another subclass instance) would show. For `C.m1` for example, I was expecting only the call inside `A.m2` and `C.m2`. Definitelly _not_ the call under `f1` or `B.m2`.

**Comment by scheglov** [comment]
@DanTup

**Comment by bwilkerson** [comment]
If I'm understanding the example correctly, I think it _is_ possible. Consider
```dart
class D extends C implements B {}
```
An instance of `D` could be passed to `f1` and the methods in `C` would be invoked.

Because of the ability to implement almost any interface, it's very difficult to prove (in general) that it's impossible to ever invoke any override of an inherited method, so we don't try.

In other words, this is working as intended, and I think we're unlikely to attempt to change it.

**Comment by FMorschel** [comment]
I see your point. But this request _does_ make sense for `final`/`sealed` classes. 

I've just tested the same code and it gives the same results. With either `C` or `B` being final there was no possible case similar to what you suggested.

I would still like for it to scan the type hierarchy to try and find a match so that case is possible to show, but my feelings are not that strong and the code where I found this behaviour could easily be changed to final classes.

Just to be clear, I would expect that result of `..m1()` to show up in either `B` or `A` and there is a way to "go to super" so in `C` we could get there.

**Comment by FMorschel** [comment]
One more thing:

Your example is flawed considering that both `B` and `C` could have gotten a method with the same name and signatures that didn't match like `int a({required num v}) => v.toInt();` and `String a(dynamic v) => v.toString();` (here the named parameter would make them not even be able to use `Never`).

This would mean that (at least today) it is impossible to have a type that would match both, but the current "See references" would still show the same calls.

**Comment by bwilkerson** [comment]
> ... this request does make sense for `final`/`sealed` classes.

Yes, certain combinations of `final` and `sealed` make sense, and we could consider adding knowledge of those modifiers to the algorithm.

> You could also say it could _in the future_ have a subclass ...

True, but a feature like this shouldn't try to reason about possible futures, only about the code that exists today.

The point of the example is that we can't definitively know whether `..m1()` might invoke the implementation in `C` without doing a whole-world analysis, and a whole-world analysis is known to be too expensive (and the analyzer, by design, never assumes that it has the whole world available to it).

We could always find ways to improve the algorithm. Some of them might be performant and some wouldn't be. That then raises the question of whether we'd end up in an uncanny valley where the imprecision was worse than if we don't try to be so precise.

**Comment by FMorschel** [comment]
> True, but a feature like this shouldn't try to reason about possible futures, only about the code that exists today.

That's more or less what your example and current implementation do here. They assume that `C` and `B` don't have conflicting declarations and show every call everywhere. 

I know the costs can be high for analysis but this can be really confusing.

I opened this issue after a talk with DanTup on the extension Discord about the following image:

![image](https://github.com/user-attachments/assets/12dc16c6-d44b-4240-9fbd-08087ca86306)

I'm looking for the references for this `inuteis` getter. It overrides a base Dao class that every other class in the right extends. I was expecting to see this class and others below it (and uses), not its "siblings". In my case, none of these has a subclass and all of the calls for `inuteis` happen inside this single file so I was really thrown back by the number of resulting references I got until I understood what was happening.

**Comment by DanTup** [comment]
I hadn't considered the point Brian made above when we first discussed this (a class using `implements`) so I agree, I think it would be wrong to filter these out. If you can possibly construct code that would be called at that location, then not considering it a reference would look like a bug.

There have been requests to VS Code/LSP to allow these kinds of results to be grouped. I don't know if that would really help here (because I'm not sure what the groups would be), but maybe if it's ever implemented we could come up with a way to split/prioritise the results in a way that makes sense? 🤔

**Comment by FMorschel** [comment]
>  we could come up with a way to split/prioritise the results in a way that makes sense?

Even though I like the idea, I don't think it would make much sense. For the splitting to make any difference for this case we would need to have the full analysis anyway so until we decide that is needed/will be implemented, it won't make much sense I don't think.

I'll leave this open for the request to consider `final`/`sealed` classes on the analysis at least. Thanks for your thoughts!

**Comment by FMorschel** [comment]
Today I found this behaviour. Repro:

```dart
abstract class Base {
  int get aaa;

  void b() {
    print(aaa);
  }
}

class Concrete extends Base {
  late int aaa;

  void c() {
    print(aaa);
  }
}
```

<img src="https://github.com/user-attachments/assets/02f9f454-07c8-4181-b515-e03bc563f498" alt="Image 1" width="50%"><img src="https://github.com/user-attachments/assets/9fb528fd-fcdb-4566-a5c2-4116f29b6f31" alt="Image 1" width="50%">

With the cursor at `aaa` inside `Base` (either the declaration or on reference inside `b` -> `print`) you get the first highlighting. But with the cursor at `aaa` inside `Concrete`, you the the second highlighting. 

This is completely consistent if you have for example:

```dart
void f(Base b, Concrete c) {
  b.aaa;
  c.aaa;
}
```

When you place the cursor under `aaa` for either `b` or `c`, it only highlights inside the respective class. This is probably because of the element being the same in all cases but different between `int get aaa;` and `late int aaa;`

---

If you rename `Concrete.aaa` to `bbb`, it also renames `Base.aaa` to the same name. This is inconsistent with the above but understabable.

---

Now with references:

Case 1:

<img src="https://github.com/user-attachments/assets/91cb0359-3953-415b-8a5d-bdecb564565d" alt="Image 1" width="100%">

Case 2:

<img src="https://github.com/user-attachments/assets/09058331-3471-4173-bbca-e5f95e0f6fac" alt="Image 1" width="100%">

This is showing yet another behaviour: the `overrides` for it don't show (but they do on `implementations` so that is understandable) but the references to all calls to it show as when you rename + the base implementation for this element.

---

Is this WAI? It feels a bit all over the place for me when we seem to have different workings for all of the above when they relate to each other so closely. Not sure if this deserves it's own issue (my point for this would be for the references and occurences/highlights to be coherent between them) but if it does please say so.

PS: On IntelliJ the only difference is that the base implementation for the element doesn't show in references, so all of the above work the same way.
